prov/util/src/util_coll.c

```diff
@@ -507,213 +503,147 @@ void util_coll_join_comp(struct util_coll_mc *coll_mc,
-	case FI_BARRIER:
-	case FI_BROADCAST:
-	case FI_ALLTOALL:
-	case FI_ALLGATHER:
-	default:
+	if (FI_MIN <= reduce_item->op && FI_BXOR >= reduce_item->op) {
+		ofi_atomic_write_handlers[reduce_item->op]
+					 [reduce_item->datatype](
+						 reduce_item->inout_buf,
+						 reduce_item->in_buf,
+						 reduce_item->count);
+	} else {
 		return -FI_ENOSYS;
-	};
+	}
+
 	return FI_SUCCESS;
 }
 
-static int util_coll_process_work_items(struct util_coll_mc *coll_mc)
-{
+int util_coll_process_pending(struct util_coll_mc *coll_mc) {
+	struct slist_entry *entry;
 	struct util_coll_hdr *hdr;
 	struct util_coll_xfer_item *xfer_item;
-	struct util_coll_reduce_item *reduce_item;
-	struct util_coll_copy_item *copy_item;
-	struct util_coll_comp_item *comp_item;
-	struct slist_entry *entry;
-	struct fi_msg_tagged msg = {0};
 	struct iovec iov;
-	int ret;
-
-	while (!slist_empty(&coll_mc->deferred_list)) {
-		entry = slist_remove_head(&coll_mc->deferred_list);
+	struct fi_msg_tagged msg;
+	int err;
+	int is_barrier;
+
+	msg.msg_iov = &iov;
+	msg.desc = NULL;
+	msg.iov_count = 1;
+	msg.ignore = 0;
+	msg.context = coll_mc;
+	msg.data = 0;
+
+	while (!slist_empty(&coll_mc->pending_xfer_list)) {
+		entry = slist_remove_head(&coll_mc->pending_xfer_list);
 		hdr = container_of(entry, struct util_coll_hdr, entry);
+		is_barrier = hdr->is_barrier;
 		switch (hdr->type) {
 		case UTIL_COLL_SEND:
-			xfer_item = (struct util_coll_xfer_item *) hdr;
+			xfer_item = container_of(
+				hdr, struct util_coll_xfer_item, hdr);
 			iov.iov_base = xfer_item->buf;
 			iov.iov_len = (xfer_item->count *
 				       ofi_datatype_size(xfer_item->datatype));
-			msg.msg_iov = &iov;
-			msg.iov_count = 1;
-			msg.addr = coll_mc->av_set->fi_addr_array[xfer_item->dest_rank];
 			msg.tag = hdr->tag;
-			msg.context = (void *) coll_mc;
-			do {
-				ret = fi_tsendmsg(coll_mc->ep, &msg, FI_COLLECTIVE);
-				if (ret && ret != -EAGAIN) {
-					slist_insert_head(&xfer_item->hdr.entry,
-							  &coll_mc->deferred_list);
-					return ret;
-				}
-			} while (ret == -EAGAIN);
-
-			slist_insert_tail(entry, &coll_mc->barrier_list);
+			msg.addr =
+				coll_mc->av_set
+					->fi_addr_array[xfer_item->dest_rank];
+			err = fi_tsendmsg(coll_mc->ep, &msg, FI_COLLECTIVE);
+			if (err) {
+				slist_insert_head(entry,
+						  &coll_mc->pending_xfer_list);
+				return err;
+			}
 			break;
 		case UTIL_COLL_RECV:
-			xfer_item = (struct util_coll_xfer_item *) hdr;
+			xfer_item = container_of(
+				hdr, struct util_coll_xfer_item, hdr);
 			iov.iov_base = xfer_item->buf;
 			iov.iov_len = (xfer_item->count *
 				       ofi_datatype_size(xfer_item->datatype));
-			msg.msg_iov = &iov;
-			msg.iov_count = 1;
-			msg.addr = coll_mc->av_set->fi_addr_array[xfer_item->src_rank];
 			msg.tag = hdr->tag;
-			msg.context = (void *) coll_mc;
-
-			ret = fi_trecvmsg(coll_mc->ep, &msg, FI_COLLECTIVE);
-			if (ret) {
-				return ret;
+			msg.addr = coll_mc->av_set
+					   ->fi_addr_array[xfer_item->src_rank];
```

Comment:
nit: unless this line is really long (>90 characters), keep as single line for better readability.